src/zino/planned_maintenance.py

```diff
@@ -0,0 +1,217 @@
+            match_expression=match_expression,
+            match_device=match_device,
+        )
+        self.planned_maintenances[pm_id] = pm
+        self._call_observers()
+        return pm
+
+    def close_planned_maintenance(self, id: int, reason: str, user: str) -> None:
+        """Deletes planned maintenance with the given id"""
+        # See `close` function in Zino 1.0 `pm.tcl`
+        pm = self.planned_maintenances.get(id, None)
+        if not pm:
+            # TODO figure out if this is enough
+            return
+        pm.add_log(f"PM closed by {user}: {reason}")
+        del self.planned_maintenances[id]
+        self._call_observers()
+
+    def get_started_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that have began since the last run of this
+        task until `now`
+        """
+        if self.last_run:
+            return [
+                pm for pm in self.planned_maintenances.values() if self.last_run < pm.start_time <= now < pm.end_time
+            ]
+        else:
+            return [pm for pm in self.planned_maintenances.values() if pm.start_time <= now < pm.end_time]
+
+    def get_ended_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that have ended since the last run of this
+        task until `now`
+        """
+        if self.last_run:
+            return [pm for pm in self.planned_maintenances.values() if self.last_run < pm.end_time <= now]
+        else:
+            return [pm for pm in self.planned_maintenances.values() if pm.end_time <= now]
+
+    def get_active_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that are currently active
+
+        This means it has started before `now` and the end_time is later than `now`
+        """
+        return [pm for pm in self.planned_maintenances.values() if pm.start_time < now < pm.end_time]
+
+    def get_old_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that should get deleted
+
+        This means that `now` is 3 days later than end_time
+        """
+        return [pm for pm in self.planned_maintenances.values() if now - pm.end_time > datetime.timedelta(days=3)]
+
+    def add_pm_observer(self, observer: PlannedMaintenanceObserver) -> None:
+        """Adds an observer function that will be called any time a planned maintenance
+        is added or removed
+        """
+        self._observers.append(observer)
+
+    def get_next_available_pm_id(self) -> int:
+        """Returns the next available planned maintenance id"""
+        self.last_pm_id += 1
+        return self.last_pm_id
+
+    def _call_observers(self) -> None:
+        for observer in self._observers:
+            observer()
+
+    def periodic(self, state: "ZinoState"):
```

Comment:
I agree this name should be more descriptive, but naming things is hard :-P

Random suggestions could be  `update_pm_calendar` or `update_pm_states`

Reply:
renamed